low-level/framework/utils/iem.py

```diff
@@ -0,0 +1,140 @@
+            ""],
+        "0050020001" : ["Kafka service is not in active state.",
+            "SSPL can not raise an alerts.",
+            "Reconfigure/start kafka service."],
+        "0050020002" : ["Kafka service is in active state.",
+            "Kafka service is available.",
+            ""]
+    }
+
+    def check_previous_iem_event(self, event_name, event_code):
+        """Before logging iem, check if is already present."""
+        previous_iem_event = None
+        is_iem_exist = False
+        if not os.path.exists(IEM_DATA_PATH):
+            os.makedirs(IEM_DATA_PATH)
+        iem_event_path = f'{IEM_DATA_PATH}/iem_{event_name}'
+        if not os.path.exists(iem_event_path):
+            with open(iem_event_path, 'w') as f:
+                f.write(event_code)
+                f.close()
+        else:
+            with open(iem_event_path, 'r') as f:
+                previous_iem_event = f.read().strip()
+                if previous_iem_event != event_code:
+                    with open(iem_event_path, 'w') as file:
+                        file.write(event_code)
+                        file.close()
+                else:
+                    logger.info("%s - IEM already created." %event_code)
+                    is_iem_exist = True
+                f.close()
+
+        return is_iem_exist
+
+    def check_fault_event(self, event_name, *events):
+        """Before logging fault_resolved iem event,
+        Check if fault iem event is present for that particular event."""
+        fault_iem = False
+        iem_event_path = f'{IEM_DATA_PATH}/iem_{event_name}'
+        if os.path.exists(iem_event_path):
+            with open(iem_event_path, 'r') as f:
+                previous_iem_event = f.read().strip()
+            if previous_iem_event in events:
+                fault_iem = True
+        return fault_iem
+
+    def generate_iem(self, severity, event_code, event_name, description):
+        """Generate iem."""
+        iem = "IEC:%sS%s:%s" %(severity, event_code, description)
+        previous_iem = self.check_previous_iem_event(event_name, event_code)
+        if not previous_iem:
+            log_iem(iem)
+
+    def create_iem_fields(self, event, severity):
+        event_code = event[0]
+        event_name = event[1]
+        description = self.EVENT_STRING[event_code][0]
+        self.generate_iem(severity, event_code, event_name, description)
+
+    def iem_fault(self, event):
+        event = self.EVENT_CODE[event]
+        severity = self.Severity["ERROR"]
+        self.create_iem_fields(event,severity)
+
+    def iem_fault_resolved(self, fault_event, fault_res_event):
+        fault_events = self.EVENT_CODE[fault_event]
+        prev_fault_iem_event = self.check_fault_event(
+        fault_events[1], fault_events[0])
+        if prev_fault_iem_event:
+            severity = self.Severity["INFO"]
+            event = self.EVENT_CODE[fault_res_event]
+            self.create_iem_fields(event, severity)
```

Comment:
How you are cleaning up fault event data, after fault resolved alert is generated?

Reply:
In case of iem_fault alert for eg, if we detect ipmitool cmd fault iem, then 
In check_previous_iem function  at first time we create file 'iem_IPMITOOL' and will add event code(fault ipmitool event code) in that file.
next when we get fault_resolved iem alert, then in same file - 'iem_IPMITOOL' , we will override prev fault event code with fault_resolved event code.